src/managers/KeybindManager.cpp

```diff
@@ -701,6 +715,16 @@ SDispatchResult CKeybindManager::handleKeybinds(const uint32_t modmask, const SP
             }
         }
 
+		if (k.longPress) {
+            m_pActiveKeybind            = &k;
+            m_pActiveKeybindEventSource = wl_event_loop_add_timer(g_pCompositor->m_sWLEventLoop, longPressHandler, &m_pActiveKeybind);
+
+            const auto PACTIVEKEEB = g_pSeatManager->keyboard.lock();
+
+            wl_event_source_timer_update(m_pActiveKeybindEventSource, PACTIVEKEEB->repeatDelay);
```

Comment:
please don't use event sources. See managers/eventLoop/EventLoopTimer.

Reply:
do you want me to include "eventLoop/EventLoopTimer.hpp"?

Reply:
ye

Reply:
shall i do it in the hpp or cpp?

Reply:
i tried:
```
            m_pActiveKeybind            = &k;

            const auto PACTIVEKEEB = g_pSeatManager->keyboard.lock();

			m_pLongPressTimer = makeShared<CEventLoopTimer>(
				std::chrono::milliseconds(PACTIVEKEEB->repeatDelay),
				longPressHandler,
				&m_pActiveKeybind
			);

			g_pEventLoopManager->addTimer(m_pLongPressTimer);
```

not sure why now it doesn't work anymore.
(m_pLongPressTimer is in the .hpp file)

Reply:
the timer should be created in the constructor, then here you'd call setTimeout instead

Reply:
im sorry that im not very familiar with the codebase, but adding the timer in the constructor and calling updateTimer later in k.longPress doesn't seem to trigger the handler(as i cant see it from the log)

Reply:
See here:
https://github.com/hyprwm/Hyprland/blob/180c26ada6233c3c59c331cb14a64aac1eeb2941/src/managers/AnimationManager.cpp#L31-L32

instead of the timeout, here we'd do `std::nullopt` instead as there is no initial time.

```cpp
m_pLongPressTimer = makeShared<CEventLoopTimer>(
				std::nullopt,
				longPressHandler,
				nullptr
			);

			g_pEventLoopManager->addTimer(m_pLongPressTimer);
```

in the ctor.

Then, here, instead:
```cpp
m_pLongPressTimer->updateTimeout(std::chrono::milliseconds(repeatDelay));
m_pLastLongPressKeybind = &k;
```

now, in the handler, just check if `m_pLastKeybind == m_pLastLongPressKeybind`

Reply:
i see. do i define the longpresshandler in the constructure too then? and what is m_pLastKeybind? and should i pass in `this` instead of nullptr since i'd need to access the p_activekeybind

Reply:
>  in the constructure too then(since it'd be out of scope otherwise)

you can throw it into a lambda, sure.

> And what is m_pLastKeybind

you used it yourself, a member of CKeybindManager. `m_pLastLongPressKeybind` is a new one I added for the sake of this functionality

> and should i pass in this instead of nullptr since i'd need to access the m_pactivekeybind

Nope! Throw it into a lambda and use an explicit `this` catch with `[this]`

Reply:
>  >And what is m_pLastKeybind

> you used it yourself, a member of CKeybindManager. m_pLastLongPressKeybind is a new one I added for the sake of this functionality

wait, you meant m_pActiveKeybind?

Reply:
Ok, after trying to directly reference the class functions in a [this] lambda func.
this is what happens,
```
			if (!m_pActiveKeybind)
				return 0;
```
result in
```
(EE) failed to read Wayland events: Broken pipe
[1]    1217210 illegal hardware instruction (core dumped)  ./build/Hyprland
```
stuff using it:
```
			SKeybind** ppActiveKeybind = (SKeybind**)m_pActiveKeybind;
        if (!*ppActiveKeybind)
return 0
```
raises SEGEV

Reply:
I need to see the entire code

Reply:
here:
```
	m_pLongPressTimer = makeShared<CEventLoopTimer>(
		std::nullopt,
		[this](SP<CEventLoopTimer> self, void* data) {
			SKeybind** ppActiveKeybind = (SKeybind**)m_pActiveKeybind;

			if (!m_pActiveKeybind) // wayland broken pipe then crash
				return 0;

			if (!*ppActiveKeybind || g_pSeatManager->keyboard.expired()) // SEGEV
				return 0;

			// const auto DISPATCHER = g_pKeybindManager->m_mDispatchers.find((*ppActiveKeybind)->handler);

			// Debug::log(LOG, "Keybind long press triggered, calling dispatcher.");
			// DISPATCHER->second((*ppActiveKeybind)->arg);
		},
		nullptr
	);

```

Reply:
you didn't add it in the ctor `g_pEventLoopManager->addTimer`

Reply:
sry, didn't specify but i did add it...

Reply:
push the code and I'll fix it

Reply:
sure, here

Reply:
remind me tomorrow if I don't fix it, I'm quite busy today but should be able to fix this in the deep evening

Reply:
ya you forgot lol. @vaxerski

Reply:
I'll do it tomorrow, super tired today, dw I have this open in a separate tab